src/metabase/measures/models/measure.clj

```diff
@@ -0,0 +1,166 @@
+(ns metabase.measures.models.measure
+  "A Measure is a saved MBQL 'macro', expanding to an `:aggregation` clause. It is tied to a table and contains
+   exactly one aggregation expression."
+  (:require
+   [clojure.set :as set]
+   [metabase.api.common :as api]
+   [metabase.lib-be.core :as lib-be]
+   [metabase.lib.core :as lib]
+   [metabase.measures.schema :as measures.schema]
+   [metabase.models.interface :as mi]
+   [metabase.models.serialization :as serdes]
+   [metabase.permissions.core :as perms]
+   [metabase.search.core :as search]
+   [metabase.util :as u]
+   [metabase.util.i18n :refer [tru]]
+   [metabase.util.log :as log]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.schema :as ms]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]))
+
+(methodical/defmethod t2/table-name :model/Measure [_model] :measure)
+(methodical/defmethod t2/model-for-automagic-hydration [:default :measure] [_original-model _k] :model/Measure)
+
+(defn- validate-mbql5-definition
+  "Validate that an MBQL 5 measure definition has the correct structure."
+  [definition]
+  (when (seq definition)
+    (mu/validate-throw ::measures.schema/measure definition)
+    definition))
+
+(defn- normalize-measure-definition
+  "Normalize measure definition.
+  Accepts:
+  - MBQL 5 full queries (passed through)
+  - MBQL 4 full queries (from serialization - converted to MBQL 5)
+  Empty seqs are normalized to `{}`."
+  [definition table-id database-id]
+  (if (seq definition)
+    (u/prog1 (-> (case (lib/normalized-mbql-version definition)
+                   (:mbql-version/mbql5 :mbql-version/legacy)
+                   definition
+                   ;; default MBQL4 fragment - wrap it in a full query
+                   {:database database-id
+                    :type :query
+                    :query (merge {:source-table table-id} definition)})
+                 lib-be/normalize-query)
+      (validate-mbql5-definition <>))
+    {}))
+
+(def ^:private transform-measure-definition
+  "Transform for measure definitions. Only handles JSON serialization/deserialization.
+  Normalization and validation happen in before-insert and after-select hooks."
+  {:in mi/json-in
+   :out mi/json-out-with-keywordization})
+
+(t2/deftransforms :model/Measure
+  {:definition transform-measure-definition})
+
```

Comment:
Duplicates my PR and copies a bunch of stuff for dealing with legacy that only applies to segments since they existed before.